### PR TITLE
Update Groovy Version and Include Swing?

### PR DESCRIPTION
**Pull request by lacan**
Hello, 

It has already happened a few times that I could benefit from having Groovy 2.5 for some scripts. 

I've also used Groovy-Swing successfully to build simple UIs in scripts rather than going the Plugin/Command way. 

Overall I think these could be interesting for more people, Swing being very optional al this point ( I can include it myself ), But Groovy 2.5 would be really nice to have :)

### CODE REVIEW

**Comment by imagejan** [comment]
Good idea to update this.

Which features from version 2.5 would you like to use? Also, I'd be very interested in seeing an example of your usage of groovy-swing.

*NB*: you could have opened a pull request directly from your `groovy-latest-swing` topic branch, instead of merging it into `master` of your own fork. Apart from creating fewer commits, this would have avoided the need of resetting your `master` branch once this PR will be merged.

**Comment by lacan** [comment]
I have so far not seen any issues or behavior that is incompatible between the two in the scripts that I have made. 

One of the nice features I liked from 2.5 is that most objects now have a 
```.digest("String algorithm")``` method that was useful to help randomize file names

I have used `tap()` to instantiate classes more easily for certain projects where it kind of made sense to pack some code into a small class rather than just a script
https://mrhaki.blogspot.com/2018/06/groovy-goodness-easy-object-creation.html

**Regarding breaking changes:** 
http://groovy-lang.org/releasenotes/groovy-2.5.html#Groovy2.5releasenotes-Breakingchanges
The only part that worries me, though I have not used it explicitly, is:
`The extension methods for the java.util.Date class are now in a separate groovy-dateutil module which isn’t included by default when using the groovy-all pom dependency.`

Regarding `groovy-swing` one example (really simple I can share is this one)

```groovy
import groovy.swing.SwingBuilder
import javax.swing.*
import java.awt.*
import ij.IJ

/*
 * GUI for selecting wells in case this is requested
 */

def sampleData = ["Item 1", "Item 2", "Item 3"]

def myGUI = new SwingBuilder()

def positionList = {
    myGUI.panel() {
        scrollPane(verticalScrollBarPolicy: JScrollPane.VERTICAL_SCROLLBAR_ALWAYS) {
            list(id: "items",
                    listData: sampleData,
                    selectionMode: ListSelectionModel.MULTIPLE_INTERVAL_SELECTION)
        } // end scrollPane
    } // end of Panel
} // end of my positions list

def myframe = myGUI.frame(title: 'Choose Items',
        location: [100, 400],
        size: [200, 300],
        defaultCloseOperation: WindowConstants.DISPOSE_ON_CLOSE) {
    panel() {
        boxLayout(axis: BoxLayout.Y_AXIS)
        label(text: 'Select multiple with Shift or Ctrlt',
                horizontalAlignment: JLabel.CENTER)

        positionList() // Add my position list

        button(text: 'Run',
                horizontalAlignment: JLabel.CENTER,
                actionPerformed: { act ->
                    def selectedItems = new HashSet(myGUI.items.getSelectedIndices().collect { val -> "Selected: " + sampleData[val] })
                    selectedItems.each { IJ.log(it) }
                    dispose() // closes the gui after the deed is done
                })
    } // end of panel
} // end of frame
// Show the GUI
myframe.setVisible(true)

```

> NB: you could have opened a pull request directly from your groovy-latest-swing
I realize this was not ideal. When I made my first pull request, I noticed later that it had been done on my `master` branch rather than the one I had forked. I can close that pull request and make a new one!

**Comment by imagejan** [comment]
> The extension methods for the java.util.Date class are now in a separate groovy-dateutil

I would then vote for also including `groovy-dateutil` as a runtime dependency. I've used the Groovy `Date()` extensions for creating timers like this one:

```groovy
timer = { description, closure ->
	// log.info("Timer started")
	timeStart = new Date()
    closure.call()
	timeStop = new Date()
	duration = TimeCategory.minus(timeStop, timeStart)
	log.info("Time spent $description: " + duration)
}

timer ("opening", { imp = IJ.openImage(filepath) })
```

---

> Regarding groovy-swing one example (really simple)

Thanks for the illustrative example.

Unfortunately, with creating custom GUIs like this one, you'll lose all the macro-recording goodness within ImageJ (let alone any headless processing via CLI or frameworks like KNIME). So even after putting your script into the ImageJ menu by saving in a `Fiji.app/scripts/` subfolder, there will be no way to run like `run("My script", "items=[Item1, Item2]")`.

While it's no problem to include the `groovy-swing` dependency, I'd very much like to see SciJava script parameters improved to cover your use case as well (which would have the benefit that you can use similar syntax from Jython, Javascript etc.). Would you mind filing an issue in `scijava-common`? Or comment on this existing one, if it covers your case: https://github.com/scijava/scijava-common/issues/260

---

> I can close that pull request and make a new one!

Don't worry, no need to create a new PR. I was just worried about the `master` branch of your fork ;-)

**Comment by lacan** [comment]
Thanks for the update!

> While it's no problem to include the groovy-swing dependency, I'd very much like to see SciJava script parameters improved to cover your use case as well (which would have the benefit that you can use similar syntax from Jython, Javascript etc.).
Thing is, this is a GUI that should only appear if the user made a particular choice ( said choice _was_ using scijava parameters). So I agree that if SciJava parameters can be invoked based on a first user decision, groovy-swing would be redundant.

File: pom.xml
Comment (imagejan) [inline pom.xml]:
```suggestion
		<groovy.version>2.5.8</groovy.version>
```

**Comment by lacan** [comment]
> I would then vote for also including groovy-dateutil as a runtime dependency

This has been added to the PR

**Review by imagejan — CHANGES_REQUESTED** [review]
I suggest using the common latest stable version `2.5.8` for all groovy artifacts.

Also, we should keep the version properties as well as the dependencies in alphabetical order (or at least all the `groovy-*` stuff together). I don't think `mvn tidy:pom` will do alphabetical sorting of dependencies, will it?

File: pom.xml
Comment (imagejan) [inline pom.xml]:
```suggestion
		<groovy-dateutil.version>${groovy.version}</groovy-dateutil.version>
```

File: pom.xml
Comment (imagejan) [inline pom.xml]:
```suggestion
		<groovy-swing.version>${groovy.version}</groovy-swing.version>
```

**Review by imagejan — CHANGES_REQUESTED** [review]
👍 

(being nitpicky: `swing` comes *before* `templates` in my alphabet ;-)

**Review by imagejan — APPROVED** [review]

**Comment by lacan** [comment]
Merci @imagejan! 

Fun interaction, sorry about the messier commit history ;)

**Comment by imagejan** [comment]
Yes, we should more often have code review within the community, shouldn't we? 🙂 

No worries about the commit history, this can all be squashed when merging.

@ctrueden any objections to merging this PR?
(In the long term, we can consider moving some property definitions to `pom-scijava`, but this shouldn't block merging this PR in my opinion.)

**Comment by ctrueden** [comment]
Thanks @lacan and @imagejan !

**Pull request merged**
